db/schema_tables.cc

```diff
@@ -1638,25 +1674,61 @@ static shared_ptr<cql3::functions::user_aggregate> create_aggregate(replica::dat
     if (initcond_str) {
         initcond = state_type->from_string(initcond_str.value());
     }
+    ::shared_ptr<cql3::functions::scalar_function> reduce_func = nullptr;
+    if (rs && !rs->empty()) {
+        auto&& rrow = rs->row(0);
+        auto rfunc = rrow.get_nonnull<sstring>("reduce_func");
+
+        reduce_func = dynamic_pointer_cast<cql3::functions::scalar_function>(
+            cql3::functions::functions::find(cql3::functions::function_name{name.keyspace, rfunc}, {state_type, state_type}));
+
+        if (!reduce_func) {
+            throw std::runtime_error(format("Reduce function {} needed by aggregate {} not found", rfunc, name.name));
+        }
+    }
+
+    return ::make_shared<cql3::functions::user_aggregate>(name, initcond, std::move(state_func), std::move(reduce_func), std::move(final_func));
+}
+
+future<lw_shared_ptr<query::result_set>> 
+extract_scylla_specific_aggregate_info(distributed<service::storage_proxy>& proxy, const query::result_set_row& row) {
+    auto aggregate_name = row.get<sstring>("aggregate_name");
+    if (!aggregate_name) {
+        co_return nullptr;
+    }
+
+    auto keyspace_name = row.get_nonnull<sstring>("keyspace_name");
+    auto arg_types = read_arg_values(row);
+    auto schema = scylla_aggregates();
 
-    return ::make_shared<cql3::functions::user_aggregate>(name, initcond, std::move(state_func), nullptr, std::move(final_func));
+    auto arg_list_type = list_type_impl::get_instance(utf8_type, false);
+    auto arg_list_val = make_list_value(arg_list_type, arg_types);
+
+    auto pkey = dht::decorate_key(*schema, partition_key::from_singular(*schema, keyspace_name));
+    auto ckey = clustering_key::from_exploded(*schema, {
+        utf8_type->decompose(aggregate_name.value()),
+        arg_list_type->decompose(arg_list_val)
+    });
 
+    co_return co_await db::system_keyspace::query(proxy, NAME, SCYLLA_AGGREGATES, pkey, query::clustering_range::make_singular(ckey));
 }
 
 static future<> merge_functions(distributed<service::storage_proxy>& proxy, schema_result before, schema_result after,
-        std::function<shared_ptr<cql3::functions::function>(replica::database& db, const query::result_set_row& row)> create) {
+        std::function<shared_ptr<cql3::functions::function>(replica::database& db, const query::result_set_row& row, lw_shared_ptr<query::result_set> rs)> create) {
     auto diff = diff_rows(before, after);
 
-    co_await proxy.local().get_db().invoke_on_all([&] (replica::database& db) {
+    co_await proxy.local().get_db().invoke_on_all([&] (replica::database& db) -> future<> {
         for (const auto& val : diff.created) {
-            cql3::functions::functions::add_function(create(db, *val));
+            auto scylla_specific_ag = co_await extract_scylla_specific_aggregate_info(proxy, *val);
```

Comment:
Why aren't you getting the row from before/after? It conflicts with how everything else works?

Reply:
I was referring to @psarna's [#10144](https://github.com/scylladb/scylla/pull/10144/commits/58529591a9ac9be49f53a69ed33cb9161406be95#diff-2cc928e139ea64f5257e11df02cd2819944fc93210107633315ee2d0edd05c27) with extracting scylla specyfic info for keyspaces.

Reply:
I don't understand.

Reply:
I also didn't include `scylla_keyspaces` table into the before/after scheme, because I considered it a little different than other system tables. While for all the tables that describe tables, keyspaces, types, and so on, their system tables actually represent the existence of something in the data dictionary. For these tables, it's very convenient to compare the diff before and after the change, because then we can distinguish if a table or a keyspace has been created, modified or deleted.

For scylla_keyspaces (and also other scylla_* tables) the situation is quite different, because it only stores auxiliary data that depends on existing items only. These tables are technically just another set of columns for existing tables, but we split them to a separate entity to avoid backward compatibility issues. And because of that, it was substantially easier to not take them into account in the already complicated `do_merge_schema` code, which would mean adding parameters to lots of functions and rpc verbs. Instead, the auxiliary information is simply loaded locally later, when we already know what was created, deleted or modified.

I expressed a similar rationale earlier when #10144 was reviewed and the series ultimately got merged, so either someone acked the idea or missed it.

Reply:
I don't understand why it's more complicated to use the same method that we used before, and less complicated to use different methods for different table.

Reply:
It makes the changes much less intrusive - for instance, `scylla_tables` was added the old way, which includes fetching and propagating an additional set of mutations, adding it to idl (https://github.com/scylladb/scylla/blob/53c66fe24ae9f25673e0d85467976177bfb2ea03/idl/frozen_schema.idl.hh#L19), and propagating this info via multiple helper functions.

With the new way, one only needs to get the additional data when generating mutations for a particular function, so the code changes are way more local - the basic interface for merging schemas stays the same, and we just generate a few more mutations only for functions that need it. The interface for generating such mutations returns vector<mutation>, so it's already designed to cover mutations from multiple tables. I think it's logically closer to what these scylla_* tables actually represent - it's just auxiliary info that would normally just be added to their corresponding original tables, but we don't do that just to stay compatible with all tools. Then, the new auxiliary info ends up serialized in the same set of mutations as everything else related to functions, and the upper layers, like idl, don't need to know that there's a separate subset of mutations that represent additional UDA information, stored in a separate optional parameter. At the same time, a cluster feature guards this auxiliary info and it will not be propagated unless the whole cluster is aware of it and capable of processing it.

Reply:
What are we losing by not propagating the mutation? I haven't analyzed it deeply but it's very surprising one method that uses RPC to propagate information is the same as another method that queries the data locally.

This is very odd.

Reply:
I still don't like it.